File: nodejs-tools/nodejs-compiler-tools/lexer-core/src/LexerInputReader.ts
import { not } from "@vighnesh153/tools";
import { LexerInput } from "./LexerInput.ts";
import { EOF_CHARACTER } from "./utils.ts";

export class LexerInputReader {
  #currentIndex = -1;
  #peekIndex = 0;
  #previousCharacter: string | EOF_CHARACTER = EOF_CHARACTER;
  #currentCharacter: string | EOF_CHARACTER = EOF_CHARACTER;

  #lineNumber = 1;
  #columnNumber = 0;

  get currentCharacter(): string | EOF_CHARACTER {
    return this.#currentCharacter;
  }

  get currentIndex(): number {
    return Math.min(this.#currentIndex, this.lexerInput.getSize());
  }

  get lineNumber(): number {
    return this.#lineNumber;
  }

  get columnNumber(): number {
    return this.#columnNumber;
  }

  constructor(private readonly lexerInput: LexerInput) {
    // Initialize the first character
    this.readNextCharacter();
  }

  readNextCharacter(): void {
    this.#previousCharacter = this.#currentCharacter;
    if (this.#peekIndex >= this.lexerInput.getSize()) {
      this.#currentCharacter = EOF_CHARACTER;
    } else {
      this.#currentCharacter = this.lexerInput.getCharacterAt(this.#peekIndex);
      this.#columnNumber++;
    }
    if (this.#previousCharacter === "\n") {
      this.#lineNumber++;
      this.#columnNumber = 1;
    }
    this.#currentIndex = this.#peekIndex;
    this.#peekIndex = Math.min(1 + this.#peekIndex, this.lexerInput.getSize());
  }

  peekCharacter(futureOffset: number = 0): string | EOF_CHARACTER {
    if (futureOffset < 0 || not(Number.isInteger(futureOffset))) {
      throw new Error(
        `Expected future offset to be a non-negative integer, found '${futureOffset}'`,
      );
    }
    const peekIndex = this.#peekIndex + futureOffset;
    if (peekIndex >= this.lexerInput.getSize()) {
      return EOF_CHARACTER;
    }
    return this.lexerInput.getCharacterAt(peekIndex);
  }
}

File: tools-jvm/XmlFormatter/src/main/kotlin/dev/vighnesh153/xml/formatter/lexer/LexerInputReader.kt
package dev.vighnesh153.xml.formatter.lexer

import kotlin.Error

class LexerInputReader(
    private val input: String,
) {
    var lineNumber = 1
        private set
    var columnNumber = 0
        private set

    var currChar: Char? = null
        private set
    private var prevChar: Char? = null

    private var currentIndex = -1
    private var peekIndex = 0

    init {
        readNextChar()
    }

    fun readNextChar() {
        prevChar = currChar
        if (peekIndex >= input.length) {
            currChar = null
        } else {
            currChar = input[peekIndex]
            columnNumber++
        }
        if (prevChar == '\n') {
            lineNumber++
            columnNumber = 1
        }
        currentIndex = peekIndex
        peekIndex = (peekIndex + 1).coerceAtMost(input.length)
    }

    fun peekChar(futureOffset: Int = 0): Char? {
        if (futureOffset < 0) {
            throw Error("Expected future offset to be non-negative, found '$futureOffset'")
        }
        val normalizedPeekIndex = peekIndex + futureOffset
        if (normalizedPeekIndex >= input.length) {
            return null
        }
        return input[normalizedPeekIndex]
    }
}